### ISSUE TITLE
FormData fails all calls - testing solution but can't make it work at all

### ISSUE THREAD

**Issue by UXDart** [jpillora/xhook#155]
Hello. using the latest npm version, any FormData usage doesn't reply, or not sure the problem. to see what happens, I downloaded zip of the repo and copied src to my project, but using

import xhook from './src/libs/xhook/main';

it seems to be doing something, but all requests are blocked. same with using .before... I can't use it in this way? I mean directly in my code? if I can make it work that way, I can see where the problem is... any tip?

Thanks

**Comment by jpillora** [comment]
have you tried using npm and `import xhook from 'xhook'` ?

@TrickyPi not sure if this is a bug or not...

**Comment by TrickyPi** [comment]
Could you provide a minimal reproducible repo? I guess the reason all requests are blocked is that your `before` function is incorrect, see [here](https://github.com/jpillora/xhook#xhookbeforehandlerrequest-callback-index)

**Comment by UXDart** [comment]
Thanks! if I use npm module directly, it works, but I have the same problem different people mentioned years ago about not working for FormData... 
to debug the issue, I tried to use the source code directly, and doesn't work at all, with the same "before" method, it was working for non FormData calls using npm. if I remove the before call also, it doesn't work (without any after call). Thanks

**Comment by UXDart** [comment]
I read before this https://github.com/jpillora/xhook/issues/152 there is a reason that using the source code in my project won't work?

**Comment by TrickyPi** [comment]
Hmmm, it works if you use source code in your project, but I recommend using `yarn link` or `npm link` to debug the source code.

**Comment by TrickyPi** [comment]
Can you provide a minimal reproducible repo or reproducible steps?

**Comment by UXDart** [comment]
found the problem, I added
import xhook from '../main';

to
xmlhttprequest.js

it is using inside xhook as it is already defined but it is not

**Comment by UXDart** [comment]
now I have another problem it seems here

https://github.com/jpillora/xhook/blob/main/src/patch/xmlhttprequest.js#L210

if I remove those lines then a call to google analytics seems to work, if I leave it setting withCredentials to false, I assume it must be true... what is the purpose of those lines?
Thanks

**Comment by TrickyPi** [comment]
The default is false, see https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials

**Comment by UXDart** [comment]
ok, but if. ""withCredentials" in xhr" is there, it means the property is there, it could have the value true... we are setting to false anyway

**Comment by UXDart** [comment]
the issue is, intercepting a call from google analytics using cors and credentials fails to connect, if I comment that line it works

**Comment by TrickyPi** [comment]
I think it can be overridden by `xhr.withCredentials=true`.

**Comment by UXDart** [comment]
yes, but what I mean, I can be wrong... if the xhr already has withCredentials true, we are setting it to false

**Comment by UXDart** [comment]
I mean, it is checking the if the property is there (field in xhr) but it is not checking if the value is true

**Comment by TrickyPi** [comment]
I think we should load "xhook" before making any request.

**Comment by UXDart** [comment]
I'm loading it before everything, will look later why it is giving an error, but google analytics can't connect and the console error says that it using cors and credentials is set to false, it seems it sending the credentials

**Comment by UXDart** [comment]
ok found the problem, here

`  facade.send = function(body) {
    //read xhr settings before hooking

    let k, modk;
    for (k of ["type", "timeout"/*, "withCredentials"*/]) {
      modk = k === "type" ? "responseType" : k;
      if (modk in facade) {
        request[k] = facade[modk];
      }
    }

    request.body = body;
    const send = function() {
`

I commented the withCredentials. the problem is that the error I was having was that credentials was set on the pre flight call to a CORS call. here it is setting the withCredentials field if it is inside facade, and it is because it set always set to false on creation. this code is before the xhr.open after that it is ok to set it
